failure on dc_ud/test_ucp_sockaddr_destroy_ep_on_err.onesided_s2c_cforce

```
[2020-05-22T11:18:10.038Z] [ RUN      ] dc_ud/test_ucp_sockaddr_destroy_ep_on_err.onesided_s2c_cforce/8 <dc_x,ud_v,ud_x,mm>
[2020-05-22T11:18:12.832Z] [     INFO ] Testing 65.65.65.12:0
[2020-05-22T11:18:12.832Z] [     INFO ] server listening on 65.65.65.12:51450
[2020-05-22T11:18:13.393Z] [r-vmb-ppc-jenkins:29862:0:29862]       flush.c:51   Fatal: ep 0x3fff9e820000: unsupported endpoint reconfiguration from 2 to 1 lanes during flush
[2020-05-22T11:18:14.784Z] 
[2020-05-22T11:18:14.784Z] /scrap/jenkins/workspace/ucx/contrib/../src/ucp/rma/flush.c: [ ucp_ep_flush_progress() ]
[2020-05-22T11:18:14.784Z]       ...
[2020-05-22T11:18:14.784Z]        46                                           req->send.flush.num_lanes;
[2020-05-22T11:18:14.784Z]        47         if (diff < 0) {
[2020-05-22T11:18:14.784Z]        48             ucs_fatal("ep %p: unsupported endpoint reconfiguration from %d to %d"
[2020-05-22T11:18:14.784Z] ==>    49                       " lanes during flush", ep, req->send.flush.num_lanes,
[2020-05-22T11:18:14.784Z]        50                       ucp_ep_num_lanes(ep));
[2020-05-22T11:18:14.784Z]        51         }
[2020-05-22T11:18:14.784Z]        52 
[2020-05-22T11:18:14.784Z] 
[2020-05-22T11:18:14.784Z] ==== backtrace (tid:  29862) ====
[2020-05-22T11:18:14.784Z]  0 0x000000000005bc58 ucs_debug_print_backtrace()  /scrap/jenkins/workspace/ucx/contrib/../src/ucs/debug/debug.c:656
[2020-05-22T11:18:14.784Z]  1 0x0000000000047718 ucp_ep_flush_progress()  /scrap/jenkins/workspace/ucx/contrib/../src/ucp/rma/flush.c:49
[2020-05-22T11:18:14.785Z]  2 0x0000000000047998 ucp_ep_flush_completion()  /scrap/jenkins/workspace/ucx/contrib/../src/ucp/rma/flush.c:236
[2020-05-22T11:18:14.785Z]  3 0x0000000000080c4c uct_invoke_completion()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/base/uct_iface.h:651
[2020-05-22T11:18:14.785Z]  4 0x0000000000080c4c uct_ud_iface_dispatch_comp()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/base/ud_inl.h:222
[2020-05-22T11:18:14.785Z]  5 0x0000000000080c4c uct_ud_ep_window_release_inline()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/base/ud_ep.c:173
[2020-05-22T11:18:14.785Z]  6 0x0000000000080c4c uct_ud_ep_process_ack()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/base/ud_ep.c:567
[2020-05-22T11:18:14.785Z]  7 0x0000000000080c4c uct_ud_ep_process_rx()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/base/ud_ep.c:777
[2020-05-22T11:18:14.785Z]  8 0x000000000008a99c uct_ud_mlx5_iface_poll_rx()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/accel/ud_mlx5.c:469
[2020-05-22T11:18:14.785Z]  9 0x000000000008a99c uct_ud_mlx5_iface_progress()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/ud/accel/ud_mlx5.c:518
[2020-05-22T11:18:14.785Z] 10 0x00000000000365d4 ucs_callbackq_dispatch()  /scrap/jenkins/workspace/ucx/contrib/../src/ucs/datastruct/callbackq.h:211
[2020-05-22T11:18:14.785Z] 11 0x00000000000365d4 uct_worker_progress()  /scrap/jenkins/workspace/ucx/contrib/../src/uct/api/uct.h:2350
[2020-05-22T11:18:14.785Z] 12 0x00000000000365d4 ucp_worker_progress()  /scrap/jenkins/workspace/ucx/contrib/../src/ucp/core/ucp_worker.c:2004
[2020-05-22T11:18:14.785Z] 13 0x000000001059841c ucp_test_base::entity::progress()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/ucp/ucp_test.cc:750
[2020-05-22T11:18:14.785Z] 14 0x0000000010598520 ucp_test::progress()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/ucp/ucp_test.cc:152
[2020-05-22T11:18:14.785Z] 15 0x000000001057e6c8 test_ucp_sockaddr::one_sided_disconnect()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:460
[2020-05-22T11:18:14.785Z] 16 0x00000000105669a8 test_ucp_sockaddr_destroy_ep_on_err_onesided_s2c_cforce_Test::test_body()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/ucp/test_ucp_sockaddr.cc:755
[2020-05-22T11:18:14.785Z] 17 0x00000000101eea08 ucs::test_base::run()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/test.cc:290
[2020-05-22T11:18:14.785Z] 18 0x00000000101eea08 ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/test.cc:316
[2020-05-22T11:18:14.785Z] 19 0x00000000103ad020 ucp_test::TestBody()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/ucp/ucp_test.h:171
[2020-05-22T11:18:14.785Z] 20 0x00000000101ca95c HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-05-22T11:18:14.785Z] 21 0x00000000101ca95c testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-05-22T11:18:14.785Z] 22 0x00000000101b9aec testing::Test::Run()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3635
[2020-05-22T11:18:14.785Z] 23 0x00000000101b9c4c testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3812
[2020-05-22T11:18:14.785Z] 24 0x00000000101b9e98 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3930
[2020-05-22T11:18:14.785Z] 25 0x00000000101bfe98 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:5808
[2020-05-22T11:18:14.785Z] 26 0x00000000101c02e4 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:5725
[2020-05-22T11:18:14.785Z] 27 0x00000000101c02e4 HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-05-22T11:18:14.785Z] 28 0x00000000101c02e4 HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-05-22T11:18:14.785Z] 29 0x00000000101c02e4 testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest-all.cc:5422
[2020-05-22T11:18:14.785Z] 30 0x00000000101448c8 RUN_ALL_TESTS()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/gtest.h:20059
[2020-05-22T11:18:14.785Z] 31 0x00000000101448c8 main()  /scrap/jenkins/workspace/ucx/contrib/../test/gtest/common/main.cc:102
[2020-05-22T11:18:14.785Z] 32 0x0000000000025100 generic_start_main.isra.0()  libc-start.c:0
[2020-05-22T11:18:14.785Z] 33 0x00000000000252f4 __libc_start_main()  ???:0
[2020-05-22T11:18:14.785Z] =================================
[2020-05-22T11:18:14.785Z] Sending notification to binl@mellanox.com
[2020-05-22T11:18:16.832Z] [r-vmb-ppc-jenkins:29862:0:29862] Process frozen...
[2020-05-22T13:32:30.217Z] make: *** [test] Terminated
```

http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/5032/pipeline/

Similar issue:
http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/8113/pipeline/
```
[2020-11-20T14:31:06.765Z] [ RUN      ] shm_ib/test_ucp_sockaddr_destroy_ep_on_err.onesided_client_sforce/4 <shm,ib/tag,cm>

[2020-11-20T14:31:07.024Z] [     INFO ] Testing 65.65.65.12:0

[2020-11-20T14:31:07.024Z] [     INFO ] server listening on 65.65.65.12:47607

[2020-11-20T14:31:17.054Z] /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:700: Failure

[2020-11-20T14:31:17.054Z] Expected: (UCS_INPROGRESS) != (status), actual: 1 vs 1

[2020-11-20T14:31:17.054Z] free not completed EP close request

[2020-11-20T14:32:25.033Z] [1605882736.837485] [r-vmb-ppc-jenkins:31167:0]      ucp_test.cc:226  UCX  ERROR request 0x10010babea8 did not complete on time

[2020-11-20T14:46:17.428Z] /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/test_helpers.cc:53: Failure

[2020-11-20T14:46:17.428Z] Failed

[2020-11-20T14:46:17.428Z] Connection timed out - abort testing

[2020-11-20T14:46:17.428Z] [r-vmb-ppc-jenkins:31167:0:31167] Caught signal 6 (Aborted: tkill(2) or tgkill(2))

[2020-11-20T14:46:17.428Z] 

[2020-11-20T14:46:17.428Z] /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc: [ ucp_test::progress() ]

[2020-11-20T14:46:17.428Z]       ...

[2020-11-20T14:46:17.428Z]       152          iter != entities().end(); ++iter)

[2020-11-20T14:46:17.428Z]       153     {

[2020-11-20T14:46:17.428Z]       154         count += (*iter)->progress(worker_index);

[2020-11-20T14:46:17.428Z] ==>   155         sched_yield();

[2020-11-20T14:46:17.428Z]       156     }

[2020-11-20T14:46:17.428Z]       157     return count;

[2020-11-20T14:46:17.428Z]       158 }

[2020-11-20T14:46:17.428Z] 

[2020-11-20T14:46:17.728Z] ==== backtrace (tid:  31167) ====

[2020-11-20T14:46:17.728Z]  0 0x000000000005ed88 ucs_debug_print_backtrace()  /scrap/jenkins/workspace/ucx-5/contrib/../src/ucs/debug/debug.c:656

[2020-11-20T14:46:17.728Z]  1 0x000000001061cc70 ucp_test::progress()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:155

[2020-11-20T14:46:17.728Z]  2 0x000000001061d46c ucp_test_base::entity::close_all_eps()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:721

[2020-11-20T14:46:17.728Z]  3 0x000000001061d604 ucp_test::disconnect()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:197

[2020-11-20T14:46:17.728Z]  4 0x000000001061d604 ucp_test_base::entity::get_num_workers()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:849

[2020-11-20T14:46:17.728Z]  5 0x000000001061d604 ucp_test::disconnect()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:187

[2020-11-20T14:46:17.728Z]  6 0x000000001061d6b0 ucp_test::cleanup()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.cc:78

[2020-11-20T14:46:17.728Z]  7 0x0000000010210140 ucs::test_base::TearDownProxy()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/test.cc:315

[2020-11-20T14:46:17.728Z]  8 0x00000000103f0e40 ucp_test::TearDown()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/ucp/ucp_test.h:190

[2020-11-20T14:46:17.728Z]  9 0x00000000101eb91c HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3562

[2020-11-20T14:46:17.728Z] 10 0x00000000101eb91c testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3598

[2020-11-20T14:46:17.728Z] 11 0x00000000101daa44 testing::Test::Run()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3643

[2020-11-20T14:46:17.728Z] 12 0x00000000101dac0c testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3812

[2020-11-20T14:46:17.728Z] 13 0x00000000101dae58 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3930

[2020-11-20T14:46:17.728Z] 14 0x00000000101e0e58 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:5808

[2020-11-20T14:46:17.728Z] 15 0x00000000101e12a4 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:5725

[2020-11-20T14:46:17.728Z] 16 0x00000000101e12a4 HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3562

[2020-11-20T14:46:17.728Z] 17 0x00000000101e12a4 HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:3598

[2020-11-20T14:46:17.728Z] 18 0x00000000101e12a4 testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest-all.cc:5422

[2020-11-20T14:46:17.728Z] 19 0x000000001015cc58 RUN_ALL_TESTS()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/gtest.h:20059

[2020-11-20T14:46:17.728Z] 20 0x000000001015cc58 main()  /scrap/jenkins/workspace/ucx-5/contrib/../test/gtest/common/main.cc:102

[2020-11-20T14:46:17.728Z] 21 0x0000000000025100 generic_start_main.isra.0()  libc-start.c:0

[2020-11-20T14:46:17.728Z] 22 0x00000000000252f4 __libc_start_main()  ???:0

[2020-11-20T14:46:17.728Z] =================================

[2020-11-20T14:46:17.728Z] Sending notification to mikhailb@nvidia.com

[2020-11-20T14:46:21.022Z] [r-vmb-ppc-jenkins:31167:0:31167] Process frozen...

[2020-11-20T16:56:52.515Z] make: *** [test] Terminated
```

--- a/src/ucp/core/ucp_ep.c
+++ b/src/ucp/core/ucp_ep.c
@@ -726,7 +726,6 @@ void ucp_ep_destroy_internal(ucp_ep_h ep)
     ucp_ep_cleanup_lanes(ep);
     if (ep->flags & UCP_EP_FLAG_INTERNAL) {
         /* it's failed tmp ep of main ep */
-        ucs_assert(ucp_ep_ext_control(ep)->local_ep_id == UCP_EP_ID_INVALID);
         ucp_ep_destroy_base(ep);
     } else {
         ucp_ep_delete(ep);
@@ -763,16 +762,36 @@ void ucp_ep_cleanup_lanes(ucp_ep_h ep)
     }
 }
 
+static int
+ucp_ep_local_disconnect_progress_remove_filter(const ucs_callbackq_elem_t *elem,
+                                               void *arg)
+{
+    ucp_ep_h ep = (ucp_ep_h)arg;
+    ucp_request_t *req;
+
+    if (elem->cb != ucp_ep_local_disconnect_progress) {
+        return 0;
+    }
+
+    req = (ucp_request_t*)elem->arg;
+    if (ep != req->send.ep) {
+        return 0;
+    }
+
+    /* we expect that only EP flush request can be remained in the callback
+     * queue, because reply UCP EP created for sending WIREUP_MSG/EP_REMOVED
+     * message is not exposed to a user */
+    ucs_assert(req->flags & UCP_REQUEST_FLAG_RELEASED);
+    ucs_assert(req->send.uct.func == ucp_ep_flush_progress_pending);
+
+    ucp_request_complete_send(req, UCS_OK);
+    return 1;
+}
+
 /* Must be called with async lock held */
 void ucp_ep_disconnected(ucp_ep_h ep, int force)
 {
-    /* remove pending slow-path progress in case it wasn't removed yet */
-    ucs_callbackq_remove_if(&ep->worker->uct->progress_q,
-                            ucp_worker_err_handle_remove_filter, ep);
-
-    /* remove pending slow-path function if it wasn't removed yet */
-    ucs_callbackq_remove_if(&ep->worker->uct->progress_q,
-                            ucp_listener_accept_cb_remove_filter, ep);
+    ucp_worker_h worker = ep->worker;
 
     ucp_ep_cm_slow_cbq_cleanup(ep);
 
@@ -791,8 +810,24 @@ void ucp_ep_disconnected(ucp_ep_h ep, int force)
         return;
     }
 
-    ucp_ep_match_remove_ep(ep->worker, ep);
+    ucp_ep_match_remove_ep(worker, ep);
     ucp_ep_destroy_internal(ep);
+
+    /* remove pending slow-path functions after EP cleanup, because it does
+     * cleanup lanes which purges all outstanding operation and purged
+     * operations could add callbacks on the progress */
+
+    /* remove pending slow-path progress in case it wasn't removed yet */
+    ucs_callbackq_remove_if(&worker->uct->progress_q,
+                            ucp_worker_err_handle_remove_filter, ep);
+
+    /* remove pending slow-path function if it wasn't removed yet */
+    ucs_callbackq_remove_if(&worker->uct->progress_q,
+                            ucp_listener_accept_cb_remove_filter, ep);
+
+    /* remove pending slow-path disconnect progress if it wasn't removed yet */
+    ucs_callbackq_remove_if(&worker->uct->progress_q,
+                            ucp_ep_local_disconnect_progress_remove_filter, ep);
 }
 
 unsigned ucp_ep_local_disconnect_progress(void *arg)
@@ -2413,15 +2448,64 @@ int ucp_ep_config_test_rndv_support(const ucp_ep_config_t *config)
            (config->key.cm_lane  != UCP_NULL_LANE);
 }
 
+/* if we have ep2iface transport we need to send an active-message based
+ * keepalive message to check the remote endpoint still exists */
 static UCS_F_ALWAYS_INLINE int
-ucp_ep_is_am_keepalive(ucp_ep_h ep, ucp_lane_index_t lane)
+ucp_ep_is_am_keepalive(ucp_ep_h ucp_ep, ucp_rsc_index_t rsc_idx)
 {
-    /* if we have ep2iface transport we need to send an active-message based
-     * keepalive message to check the remote endpoint still exists */
-    return (ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* remote ID defined */
-           (ep->flags & UCP_EP_FLAG_REMOTE_CONNECTED) && /* wireup completed */
-           (ucp_ep_get_iface_attr(ep, lane)->cap.flags &
-            UCT_IFACE_FLAG_CONNECT_TO_IFACE); /* connect-to-iface */
+    ucp_worker_iface_t *wiface;
+
+    if (!(ucp_ep->flags & UCP_EP_FLAG_REMOTE_ID) ||
+        (rsc_idx == UCP_NULL_RESOURCE)) {
+        /* if remote ID isn't defined or rsc index is NULL (i.e. it is CM lane),
+         * don't do AM keepalive */
+        return 0;
+    }
+
+    wiface = ucp_worker_iface(ucp_ep->worker, rsc_idx);
+    return ucs_test_all_flags(wiface->attr.cap.flags,
+                              UCT_IFACE_FLAG_CONNECT_TO_IFACE |
+                              UCT_IFACE_FLAG_AM_BCOPY);
+}
+
+ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
+                                        ucp_rsc_index_t rsc_idx, unsigned flags,
+                                        uct_completion_t *comp)
+{
+    ucp_tl_bitmap_t tl_bitmap = UCS_BITMAP_ZERO;
+    ucs_status_t status;
+    ssize_t packed_len;
+    struct iovec wireup_msg_iov[2];
+    ucp_wireup_msg_t wireup_msg;
+
+    ucs_assert((rsc_idx == UCP_NULL_RESOURCE) ||
+               (ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
+                UCT_IFACE_FLAG_EP_CHECK));
+
+    if (!ucp_ep_is_am_keepalive(ucp_ep, rsc_idx)) {
+        return uct_ep_check(uct_ep, flags, comp);
+    }
+
+    ucs_assert(ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
+               UCT_IFACE_FLAG_AM_BCOPY);
+
+    UCS_BITMAP_SET(tl_bitmap, rsc_idx);
+
+    status = ucp_wireup_msg_prepare(ucp_ep, UCP_WIREUP_MSG_EP_CHECK,
+                                    &tl_bitmap, NULL, &wireup_msg,
+                                    &wireup_msg_iov[1].iov_base,
+                                    &wireup_msg_iov[1].iov_len);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    wireup_msg_iov[0].iov_base = &wireup_msg;
+    wireup_msg_iov[0].iov_len  = sizeof(wireup_msg);
+
+    packed_len = uct_ep_am_bcopy(uct_ep, UCP_AM_ID_WIREUP,
+                                 ucp_wireup_msg_pack, wireup_msg_iov, 0);
+    ucs_free(wireup_msg_iov[1].iov_base);
+    return (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
 }
 
 void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
@@ -2430,9 +2514,8 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
     ucp_lane_index_t lane;
     ucs_status_t status;
     ucp_rsc_index_t rsc_index;
-    ucp_tl_bitmap_t tl_bitmap;
 
-    if (ep->flags & (UCP_EP_FLAG_FAILED | UCP_EP_FLAG_CLOSED)) {
+    if (ep->flags & UCP_EP_FLAG_FAILED) {
         *lane_map = 0;
         return;
     }
@@ -2441,29 +2524,22 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
     check_lanes = *lane_map & ucp_ep_config(ep)->key.ep_check_map;
 
     ucs_for_each_bit(lane, check_lanes) {
-        UCS_BITMAP_CLEAR(&tl_bitmap);
         ucs_assert(lane < UCP_MAX_LANES);
-        /* in case if remote ID is defined, wireup is completed and EP is
-         * in connect-to-iface mode then use UCP/AM keepalive */
-        if (ucp_ep_is_am_keepalive(ep, lane)) {
-            rsc_index = ucp_ep_get_rsc_index(ep, lane);
-            UCS_BITMAP_SET(tl_bitmap, rsc_index);
-            status    = ucp_wireup_msg_send(ep, UCP_WIREUP_MSG_EP_CHECK,
-                                            &tl_bitmap, NULL);
-            ucs_assert(status == UCS_OK);
-            *lane_map &= ~UCS_BIT(lane);
-        } else {
-            /* coverity[overrun-local] */
-            status = uct_ep_check(ep->uct_eps[lane], 0, NULL);
-            if (status == UCS_OK) {
-                *lane_map &= ~UCS_BIT(lane);
-            } else if (status != UCS_ERR_NO_RESOURCE) {
-                *lane_map &= ~UCS_BIT(lane);
-                ucs_warn("unexpected return status from uct_ep_check(ep=%p, "
-                         "lane[%d]=%p): %s",
-                         ep, lane, ep->uct_eps[lane],
-                         ucs_status_string(status));
-            }
+        rsc_index = ucp_ep_get_rsc_index(ep, lane);
+        ucs_assert((rsc_index != UCP_NULL_RESOURCE) ||
+                   (lane == ucp_ep_get_cm_lane(ep)));
+
+        status = ucp_ep_do_uct_ep_keepalive(ep, ep->uct_eps[lane], rsc_index, 0,
+                                            NULL);
+        if (status == UCS_ERR_NO_RESOURCE) {
+            continue;
+        } else if (status != UCS_OK) {
+            ucs_warn("unexpected return status from doing keepalive(ep=%p, "
+                     "lane[%d]=%p): %s",
+                     ep, lane, ep->uct_eps[lane],
+                     ucs_status_string(status));
         }
+
+        *lane_map &= ~UCS_BIT(lane);
     }
 }

--- a/src/ucp/core/ucp_ep.h
+++ b/src/ucp/core/ucp_ep.h
@@ -585,6 +585,8 @@ void ucp_ep_invoke_err_cb(ucp_ep_h ep, ucs_status_t status);
 
 int ucp_ep_config_test_rndv_support(const ucp_ep_config_t *config);
 
+ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self);
+
 void ucp_ep_flush_completion(uct_completion_t *self);
 
 void ucp_ep_flush_request_ff(ucp_request_t *req, ucs_status_t status);
@@ -594,6 +596,21 @@ void ucp_ep_discard_lanes(ucp_ep_h ucp_ep, ucs_status_t status);
 void ucp_ep_register_disconnect_progress(ucp_request_t *req);
 
 ucp_lane_index_t ucp_ep_lookup_lane(ucp_ep_h ucp_ep, uct_ep_h uct_ep);
+
+/**
+ * @brief Do keepalive operation for a specific UCT EP.
+ *
+ * @param [in] ucp_ep  UCP Endpoint object to operate keepalive.
+ * @param [in] uct_ep  UCT Endpoint object to do keepalive on.
+ * @param [in] rsc_idx Resource index to check. 
+ * @param [in] flags   Flags for keepalive operation.
+ * @param [in] comp    Pointer to keepalive completion object.
+ *
+ * @return Status of keepalive operation.
+ */
+ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
+                                        ucp_rsc_index_t rsc_idx, unsigned flags,
+                                        uct_completion_t *comp);
 
 /**
  * @brief Do keepalive operation.

--- a/src/ucp/rma/flush.c
+++ b/src/ucp/rma/flush.c
@@ -201,7 +201,7 @@ static unsigned ucp_ep_flush_resume_slow_path_callback(void *arg)
     return 0;
 }
 
-static ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
+ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
     ucp_lane_index_t lane = req->send.lane;

--- a/src/ucp/wireup/wireup.c
+++ b/src/ucp/wireup/wireup.c
@@ -22,6 +22,7 @@
 #include <ucp/tag/eager.h>
 #include <ucs/async/async.h>
 #include <ucs/datastruct/queue.h>
+#include <ucs/sys/iovec.h>
 
 /*
  * Description of the protocol in UCX wiki:
@@ -44,12 +45,13 @@
     } while (0)
 
 
-static size_t ucp_wireup_msg_pack(void *dest, void *arg)
+size_t ucp_wireup_msg_pack(void *dest, void *arg)
 {
-    ucp_request_t *req = arg;
-    *(ucp_wireup_msg_t*)dest = req->send.wireup;
-    memcpy((ucp_wireup_msg_t*)dest + 1, req->send.buffer, req->send.length);
-    return sizeof(ucp_wireup_msg_t) + req->send.length;
+    struct iovec *wireup_msg_iov = (struct iovec*)arg;
+
+    return ucs_iov_copy(wireup_msg_iov, 2, 0, dest,
+                        wireup_msg_iov[0].iov_len + wireup_msg_iov[1].iov_len,
+                        UCS_IOV_COPY_TO_BUF);
 }
 
 static const char* ucp_wireup_msg_str(uint8_t msg_type)
@@ -102,9 +104,10 @@ ucs_status_t ucp_wireup_msg_progress(uct_pending_req_t *self)
 {
     ucp_request_t *req  = ucs_container_of(self, ucp_request_t, send.uct);
     ucp_ep_h ep         = req->send.ep;
-    ucs_status_t status = UCS_OK;
+    ucs_status_t status;
     ssize_t packed_len;
     unsigned am_flags;
+    struct iovec wireup_msg_iov[2];
 
     UCS_ASYNC_BLOCK(&ep->worker->async);
 
@@ -112,6 +115,7 @@ ucs_status_t ucp_wireup_msg_progress(uct_pending_req_t *self)
         if (ep->flags & UCP_EP_FLAG_REMOTE_CONNECTED) {
             ucs_trace("ep %p: not sending wireup message - remote already connected",
                       ep);
+            status = UCS_OK;
             goto out;
         }
     } else if (req->send.wireup.type == UCP_WIREUP_MSG_PRE_REQUEST) {
@@ -130,8 +134,14 @@ ucs_status_t ucp_wireup_msg_progress(uct_pending_req_t *self)
     VALGRIND_CHECK_MEM_IS_DEFINED(&req->send.wireup, sizeof(req->send.wireup));
     VALGRIND_CHECK_MEM_IS_DEFINED(req->send.buffer, req->send.length);
 
+    wireup_msg_iov[0].iov_base = &req->send.wireup;
+    wireup_msg_iov[0].iov_len  = sizeof(req->send.wireup);
+
+    wireup_msg_iov[1].iov_base = req->send.buffer;
+    wireup_msg_iov[1].iov_len  = req->send.length;
+
     packed_len = uct_ep_am_bcopy(ep->uct_eps[req->send.lane], UCP_AM_ID_WIREUP,
-                                 ucp_wireup_msg_pack, req, am_flags);
+                                 ucp_wireup_msg_pack, wireup_msg_iov, am_flags);
     if (ucs_unlikely(packed_len < 0)) {
         status = (ucs_status_t)packed_len;
         if (ucs_likely(status == UCS_ERR_NO_RESOURCE)) {
@@ -168,16 +178,46 @@ out:
     return status;
 }
 
+ucs_status_t
+ucp_wireup_msg_prepare(ucp_ep_h ep, uint8_t type,
+                       const ucp_tl_bitmap_t *tl_bitmap,
+                       const ucp_lane_index_t *lanes2remote,
+                       ucp_wireup_msg_t *msg_hdr, void **address_p,
+                       size_t *address_length_p)
+{
+    ucs_status_t status;
+
+    msg_hdr->type      = type;
+    msg_hdr->err_mode  = ucp_ep_config(ep)->key.err_mode;
+    msg_hdr->conn_sn   = ep->conn_sn;
+    msg_hdr->src_ep_id = ucp_ep_local_id(ep);
+    if (ep->flags & UCP_EP_FLAG_REMOTE_ID) {
+        msg_hdr->dst_ep_id = ucp_ep_remote_id(ep);
+    } else {
+        msg_hdr->dst_ep_id = UCP_EP_ID_INVALID;
+    }
+
+    /* pack all addresses */
+    status = ucp_address_pack(ep->worker, ep, tl_bitmap,
+                              UCP_ADDRESS_PACK_FLAGS_ALL, lanes2remote,
+                              address_length_p, address_p);
+    if (status != UCS_OK) {
+        ucs_error("failed to pack address: %s", ucs_status_string(status));
+    }
+
+    return status;
+}
+
 /*
  * @param [in] rsc_tli  Resource index for every lane.
  */
-ucs_status_t ucp_wireup_msg_send(ucp_ep_h ep, uint8_t type,
-                                 const ucp_tl_bitmap_t *tl_bitmap,
-                                 const ucp_lane_index_t *lanes2remote)
+
+static ucs_status_t
+ucp_wireup_msg_send(ucp_ep_h ep, uint8_t type, const ucp_tl_bitmap_t *tl_bitmap,
+                    const ucp_lane_index_t *lanes2remote)
 {
-    ucp_request_t* req;
+    ucp_request_t *req;
     ucs_status_t status;
-    void *address;
 
     ucs_assert(ep->cfg_index != UCP_WORKER_CFG_INDEX_NULL);
 
@@ -186,36 +226,23 @@ ucs_status_t ucp_wireup_msg_send(ucp_ep_h ep, uint8_t type,
      */
     req = ucs_malloc(sizeof(*req), "wireup_msg_req");
     if (req == NULL) {
+        ucs_error("failed to allocate request for sending WIREUP message");
         return UCS_ERR_NO_MEMORY;
     }
 
-    req->flags                   = 0;
-    req->send.ep                 = ep;
-    req->send.wireup.type        = type;
-    req->send.wireup.err_mode    = ucp_ep_config(ep)->key.err_mode;
-    req->send.wireup.conn_sn     = ep->conn_sn;
-    req->send.wireup.src_ep_id   = ucp_ep_local_id(ep);
-    if (ep->flags & UCP_EP_FLAG_REMOTE_ID) {
-        req->send.wireup.dst_ep_id = ucp_ep_remote_id(ep);
-    } else {
-        req->send.wireup.dst_ep_id = UCP_EP_ID_INVALID;
-    }
-
-    req->send.uct.func           = ucp_wireup_msg_progress;
-    req->send.datatype           = ucp_dt_make_contig(1);
+    req->flags         = 0;
+    req->send.ep       = ep;
+    req->send.uct.func = ucp_wireup_msg_progress;
+    req->send.datatype = ucp_dt_make_contig(1);
     ucp_request_send_state_init(req, ucp_dt_make_contig(1), 0);
 
-    /* pack all addresses */
-    status = ucp_address_pack(ep->worker, ep, tl_bitmap,
-                              UCP_ADDRESS_PACK_FLAGS_ALL, lanes2remote,
-                              &req->send.length, &address);
+    status = ucp_wireup_msg_prepare(ep, type, tl_bitmap, lanes2remote,
+                                    &req->send.wireup, &req->send.buffer,
+                                    &req->send.length);
     if (status != UCS_OK) {
         ucs_free(req);
-        ucs_error("failed to pack address: %s", ucs_status_string(status));
         return status;
     }
-
-    req->send.buffer = address;
 
     ucp_request_send(req, 0);
     return UCS_OK;

--- a/src/ucp/wireup/wireup.h
+++ b/src/ucp/wireup/wireup.h
@@ -90,9 +90,6 @@ typedef struct {
 } ucp_wireup_select_info_t;
 
 
-ucs_status_t ucp_wireup_msg_send(ucp_ep_h ep, uint8_t type, const ucp_tl_bitmap_t *tl_bitmap,
-                                 const ucp_lane_index_t *lanes2remote);
-
 ucs_status_t ucp_wireup_send_request(ucp_ep_h ep);
 
 ucs_status_t ucp_wireup_send_pre_request(ucp_ep_h ep);
@@ -110,7 +107,16 @@ double ucp_wireup_amo_score_func(ucp_context_h context,
                                  const uct_iface_attr_t *iface_attr,
                                  const ucp_address_iface_attr_t *remote_iface_attr);
 
+size_t ucp_wireup_msg_pack(void *dest, void *arg);
+
 ucs_status_t ucp_wireup_msg_progress(uct_pending_req_t *self);
+
+ucs_status_t
+ucp_wireup_msg_prepare(ucp_ep_h ep, uint8_t type,
+                       const ucp_tl_bitmap_t *tl_bitmap,
+                       const ucp_lane_index_t *lanes2remote,
+                       ucp_wireup_msg_t *msg_hdr, void **address_p,
+                       size_t *address_length_p);
 
 int ucp_wireup_msg_ack_cb_pred(const ucs_callbackq_elem_t *elem, void *arg);
 

--- a/src/ucp/wireup/wireup_cm.c
+++ b/src/ucp/wireup/wireup_cm.c
@@ -331,6 +331,9 @@ static ssize_t ucp_cm_client_priv_pack_cb(void *arg,
         goto out_check_err;
     }
 
+    ucp_ep_ext_control(cm_wireup_ep->tmp_ep)->local_ep_id =
+            ucp_ep_ext_control(ep)->local_ep_id;
+
     cm_wireup_ep->tmp_ep->flags |= UCP_EP_FLAG_INTERNAL;
     ucs_debug("ep %p: created tmp_ep %p", ep, cm_wireup_ep->tmp_ep);
 

--- a/src/ucp/wireup/wireup_ep.c
+++ b/src/ucp/wireup/wireup_ep.c
@@ -332,33 +332,46 @@ static ucs_status_t ucp_wireup_ep_flush(uct_ep_h uct_ep, unsigned flags,
     return UCS_ERR_NO_RESOURCE;
 }
 
+static ucs_status_t
+ucp_wireup_ep_do_check(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
+                       ucp_rsc_index_t rsc_idx, unsigned flags,
+                       uct_completion_t *comp)
+{
+    ucp_worker_h worker = ucp_ep->worker;
+    ucp_worker_iface_t *wiface;
+
+    ucs_assert(rsc_idx != UCP_NULL_RESOURCE);
+
+    wiface = ucp_worker_iface(worker, rsc_idx);
+    if (wiface->attr.cap.flags & UCT_IFACE_FLAG_EP_CHECK) {
+        return ucp_ep_do_uct_ep_keepalive(ucp_ep, uct_ep, rsc_idx, flags,
+                                          comp);
+    }
+
+    /* if EP_CHECK is not supported by UCT transport, it has to support a
+     * built-in keepalive mechanism to be able to detect peer failure during
+     * wireup
+     */
+    ucs_assert(wiface->attr.cap.flags & UCT_IFACE_FLAG_EP_KEEPALIVE);
+    return UCS_OK;
+}
+
 static ucs_status_t ucp_wireup_ep_check(uct_ep_h uct_ep, unsigned flags,
                                         uct_completion_t *comp)
 {
     ucp_wireup_ep_t *wireup_ep = ucp_wireup_ep(uct_ep);
-    ucp_worker_h worker        = wireup_ep->super.ucp_ep->worker;
+    ucp_ep_h ucp_ep            = wireup_ep->super.ucp_ep;
     ucp_ep_h tmp_ep            = wireup_ep->tmp_ep;
     ucp_lane_index_t lane;
-    ucp_worker_iface_t *wiface;
 
     if (wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY) {
         return uct_ep_check(wireup_ep->super.uct_ep, flags, comp);
     }
 
     if (wireup_ep->aux_ep != NULL) {
-        ucs_assert(wireup_ep->aux_rsc_index != UCP_NULL_RESOURCE);
-
-        wiface = ucp_worker_iface(worker, wireup_ep->aux_rsc_index);
-        if (wiface->attr.cap.flags & UCT_IFACE_FLAG_EP_CHECK) {
-            return uct_ep_check(wireup_ep->aux_ep, flags, comp);
-        }
-
-        /* if EP_CHECK is not supported by auxiliary transport, it has to
-         * support a built-in keepalive mechanism to be able to detect peer
-         * failure during wireup
-         */
-        ucs_assert(wiface->attr.cap.flags & UCT_IFACE_FLAG_EP_KEEPALIVE);
-        return UCS_OK;
+        return ucp_wireup_ep_do_check(ucp_ep, wireup_ep->aux_ep,
+                                      wireup_ep->aux_rsc_index,
+                                      flags, comp);
     }
 
     if ((tmp_ep != NULL) && (ucp_ep_config(tmp_ep)->key.ep_check_map != 0)) {
@@ -373,7 +386,10 @@ static ucs_status_t ucp_wireup_ep_check(uct_ep_h uct_ep, unsigned flags,
         }
 
         wireup_ep->tmp_ep_check_map &= ~UCS_BIT(lane);
-        return uct_ep_check(tmp_ep->uct_eps[lane], flags, comp);
+        ucs_assert(ucp_ep_remote_id(tmp_ep) == ucp_ep_remote_id(ucp_ep));
+        return ucp_wireup_ep_do_check(tmp_ep, tmp_ep->uct_eps[lane],
+                                      ucp_ep_get_rsc_index(tmp_ep, lane),
+                                      flags, comp);
     }
 
     return UCS_OK;

